c/storageApiml.c

```diff
@@ -0,0 +1,648 @@
+    .method = "DELETE",
+    .path = path,
+    .body = NULL,
+    .bodyLen = 0,
+    .parseJson = false
+  };
+  ApimlResponse *response = apimlDoRequest(storage, &request, &status);
+  if (status) {
+    *statusOut = status;
+    return;
+  }
+  int statusCode = response->statusCode;
+  *statusOut = transformHttpStatus(statusCode);
+  freeApimlResponse(response);
+  zowelog(NULL, LOG_COMP_ID_APIML_STORAGE, ZOWE_LOG_DEBUG, "http response status %d\n", statusCode);
+}
+
+static void apimlStorageSetString(ApimlStorage *storage, const char *key, const char *value, int *statusOut) {
+  int status = 0;
+  createOrChange(storage, OP_CHANGE, key, value, &status);
+  if (status == STORAGE_STATUS_KEY_NOT_FOUND) {
+    createOrChange(storage, OP_CREATE, key, value, &status);
+  }
+  *statusOut = status;
+}
+
+static const char *MESSAGES[] = {
+  [STORAGE_STATUS_HTTP_ERROR] = "Failed to send HTTP request",
+  [STORAGE_STATUS_RESPONSE_ERROR] = "Error receiving response",
+  [STORAGE_STATUS_JSON_RESPONSE_ERROR] = "Error parsing JSON response",
+  [STORAGE_STATUS_ALLOC_ERROR] = "Failed to allocate memory",
+  [STORAGE_STATUS_INVALID_KV_RESPONSE] = "Invalid key/value response",
+  [STORAGE_STATUS_INVALID_CLIENT_CERT] = "Invalid client certificate",
+};
+
+#define MESSAGE_COUNT sizeof(MESSAGES)/sizeof(MESSAGES[0])
+
+static const char *apimlStorageGetStrStatus(ApimlStorage *storage, int status) {
+  if (status >= MESSAGE_COUNT || status < STORAGE_STATUS_FIRST_CUSTOM_STATUS) {
+    return "Unknown status code";
+  }
+  const char *message = MESSAGES[status];
+  if (!message) {
+    return "Unknown status code";
+  }
+  return message;
+}
+
+static makeKeyPrefix(const char *pluginId, char *keyBuffer, int keyBufferSize) {
+  int pluginIdLen = strlen(pluginId);
+  int pos = 0;
+  for (int i = 0; i < pluginIdLen; i++) {
+    // Currently, only alpha-numeric chars allowed in keys
+    // TODO: Use pluginId as a prefix when https://github.com/zowe/api-layer/issues/1300 has resolved
+    if (isalnum(pluginId[i]) && pos < keyBufferSize - 1) {
+      keyBuffer[pos++] = pluginId[i];
+    }
+  }
+  keyBuffer[pos] = '\0';
+}
+
+Storage *makeApimlStorage(ApimlStorageSettings *settings, const char *pluginId) {
+  if (!settings) {
+    return NULL;
+  }
+  Storage *storage = (Storage*)safeMalloc(sizeof(*storage), "Storage");
+  if (!storage) {
+    return NULL;
+  }
+  ApimlStorage *apimlStorage = (ApimlStorage*)safeMalloc(sizeof(*apimlStorage), "APIML Storage");
+  if (!apimlStorage) {
+    safeFree((char*)storage, sizeof(*storage));
+    return NULL;
+  }
+  HttpClientSettings *clientSettings = (HttpClientSettings*)safeMalloc(sizeof(*clientSettings), "HTTP Client Settings");
+  if (!clientSettings) {
+    safeFree((char*)storage, sizeof(*storage));
+    safeFree((char*)apimlStorage, sizeof(*apimlStorage));
+    return NULL;
+  }
+  clientSettings->host = settings->host;
+  clientSettings->port = settings->port;
+  clientSettings->recvTimeoutSeconds = 10;
```

Comment:
10 seconds should be enough, but shouldnt this be a parameter from somewhere?

Reply:
I moved the parameter into `APIMLStorageSettings` so that we can make it configurable via some `ZWED_` environment variable or using `server.json` if we need it in the future.